apps/scratch/src/components/block-config/BlockConfig.tsx

```diff
@@ -0,0 +1,159 @@
+import { FormEvent, useCallback, useEffect, useState } from "react";
+import { ModifyBlockConfigEvent } from "../../events/modify-block-config";
+
+import styles from "./block-config.css";
+import VM from "scratch-vm";
+import { UpdateBlockToolboxEvent } from "../../events/update-block-toolbox";
+
+const cannotBeUsed = 0;
+const infiniteUses = -1;
+
+const BlockConfig = ({ vm }: { vm: VM }) => {
+  const [blockId, setBlockId] = useState<string | null>(null);
+  const [canBeUsed, setCanBeUsed] = useState<boolean>(true);
+  const [hasBlockLimit, setHasBlockLimit] = useState<boolean>(true);
+  const [blockLimit, setBlockLimit] = useState<string>("1");
+
+  const onModifyBlockConfig = useCallback(
+    (e: Event) => {
+      const config = vm.crtConfig;
+      if (!(e instanceof ModifyBlockConfigEvent) || !config) {
+        return;
+      }
+
+      const currentConfig = config.allowedBlocks[e.blockId];
+
+      if (typeof currentConfig === "boolean") {
+        console.error(
+          `Trying to limit the number of blocks for ${e.blockId} which is not configured for this`,
+        );
+        return;
+      }
+
+      const blockCanBeUsed =
+        currentConfig !== undefined && currentConfig !== cannotBeUsed;
+
+      setCanBeUsed(blockCanBeUsed);
+      setHasBlockLimit(blockCanBeUsed && currentConfig !== infiniteUses);
+
+      const limit = currentConfig || 1;
+
+      setBlockLimit(limit.toString());
+      setBlockId(e.blockId);
+    },
+    [vm],
+  );
+
+  useEffect(() => {
+    window.addEventListener(
+      ModifyBlockConfigEvent.eventName,
+      onModifyBlockConfig,
+    );
+
+    return () => {
+      window.removeEventListener(
+        ModifyBlockConfigEvent.eventName,
+        onModifyBlockConfig,
+      );
+    };
+  }, [onModifyBlockConfig]);
+
+  const onSubmit = useCallback(
+    (e: FormEvent<HTMLFormElement>) => {
+      e.preventDefault();
+      e.stopPropagation();
+
+      const config = vm.crtConfig;
+
+      if (!config) {
+        console.error("No task config found");
+        return;
+      }
+
+      if (!blockId) {
+        console.error("Block ID not found", blockId);
+        return;
+      }
+
+      const limitOrNaN = parseInt(blockLimit, 10);
+      const limit = isNaN(limitOrNaN) ? 1 : limitOrNaN;
+
+      // -1 means infinite, 0 means not allowed
+      config.allowedBlocks[blockId] = canBeUsed
+        ? hasBlockLimit
+          ? limit
+          : infiniteUses
+        : cannotBeUsed;
```

Comment:
nested ternary operators are a bit unclear though, you might want to do the following for readability
```suggestion
      config.allowedBlocks[blockId] = canBeUsed && hasBlockLimit ? limit
        : canBeUsed ? infiniteUses
        : cannotBeUsed;
```